### PR TITLE
#2599 Fix InfoWindow position after style switching

### PR DESCRIPTION
**Pull request by ljbade**
Fixes InfoWindow position after style switching (for #2599)

Also incorporates the #2588 commit that @tobrun reverted with the crash fixed.

As part of this #630 was implemented.

### CODE REVIEW

**Comment by ljbade** [comment]
@jfirebaugh Can you review the changes made to core for implementing #630 

@tobrun Can you verify the behaviour of InfoWindow and InfoWindowAdapter. Try changing the style to Emerald and back to see how it now readjusts the InfoWindow position.

**Comment by ljbade** [comment]
@tobrun I think there is a bug with InfoWindow and InfoWindowAdapter activities. It does not seem to load the offset.

Going to need to workout why the map finish loading event is not updating the markers on these activities.

**Comment by tobrun** [comment]
@ljbade 
When testing I noticed that the MainActivity did not give any problems for adding markers/infowindow on a long press. I'm thinking it could be related to adding markers/infowindow while creating the MapView?

**Comment by ljbade** [comment]
@tobrun Yeah it has something to do with adding Marker during onCreate.

**Comment by ljbade** [comment]
@jfirebaugh How do I move that notification back to the main thread, since `setStyleUrl` is async? Is there a function I can use to post an event to the main thread?

**Comment by ljbade** [comment]
@jfirebaugh or should I check for failure in `renderSync`?

**Comment by jfirebaugh** [comment]
I don't think there's a good way to emit `MapChangeDidFailLoadingMap` right now.

**Comment by ljbade** [comment]
@jfirebaugh An idea I had was to add `loadFailed` flag to `MapData` then set this on fail. `renderSync` will then check and clear this flag before it checks the `loading` flag.

**Comment by ljbade** [comment]
I had to move the set for `data.loading` to `loadStyleJSON()` to avoid a false map loaded event right after the map is created (and finishes rendering a blank map) but before it starts loading the new style.

**Comment by ljbade** [comment]
Only Travis failure is a known iOS hang. Other jobs passed.

File: src/mbgl/map/map_context.cpp [hunk 1]
```diff
@@ -92,10 +92,6 @@ void MapContext::triggerUpdate(const TransformState& state, const Update flags)
 }
 
 void MapContext::setStyleURL(const std::string& url) {
-    if (styleURL == url) {
```
Comment (jfirebaugh) [inline src/mbgl/map/map_context.cpp]: You're reverting 074d61a55e32e83173ec12b2eb34b78ddaaf0ef1 from earlier today. Accidental?

**Comment by ljbade** [comment]
@jfirebaugh I missed you comment before merging. I used `git cherry-pick 074d61a55e32e83173ec12b2eb34b78ddaaf0ef1` to replay @brunoabinader's commit on master to restore it.

**Pull request merged**